### PR TITLE
[GPU] add blocked format to experimental detectron detection output

### PR DESCRIPTION
**Pull request by OlehKravchyshyn**
### Details:
 - added blocked format to experimental detectron detection output
 - added tests for testing blocked format

### Tickets:
 - 89796

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/eddo/experimental_detectron_detection_output_kernel_ref.cpp [hunk 1]
```diff
@@ -21,10 +21,12 @@ ParamsKey ExperimentalDetectronDetectionOutputKernelRef::GetSupportedKey() const
     k.EnableOutputDataType(Datatype::F32);
     k.EnableOutputDataType(Datatype::INT32);
     k.EnableOutputDataType(Datatype::INT64);
-    k.EnableInputLayout(DataLayout::bfyx);
-    k.EnableOutputLayout(DataLayout::bfyx);
+
+    k.EnableAllInputLayout();
+    k.EnableAllOutputLayout();
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/actual_kernels/eddo/experimental_detectron_detection_output_kernel_ref.cpp:26]: All inputs and outputs maybe not very good idea as only 4D is supported

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl [hunk 2]
```diff
@@ -133,21 +133,28 @@ KERNEL(eddo_ref_stage_0)
...
-
         // width & height of box
         INPUT_TYPE2 box_size = (box.hi - box.lo + COORDINATE_OFFSET);
 
         // center location of box
         const INPUT_TYPE2 center = box.lo + HALF_ONE * box_size;
 
-        const INPUT_TYPE4 delta = vload4(offset, deltas) / DELTA_WEIGHTS;
+        INPUT_TYPE4 delta;
+        delta[0] = deltas[INPUT1_GET_INDEX(roi_idx, class_idx * 4, 0, 0)];
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl:153]: I'm not sure for performance, but maybe better to precalculate class_idx*4   (and don't now about opencl optimizer- maybe better to use class_idx<<2)

**Comment by wilson-seok** [comment]
No performance regression during int8 models.

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Build error happened in build_rhel82_release. Please check it.
image

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Build error happened in build_ubuntu18_release also. Please check it.
image

**Comment by tgubanova-lohika** [comment]
> @OlehKravchyshyn Build error happened in build_rhel82_release. Please check it. image

@wilson-seok As I mentioned in mail this PR is connected with  reverting  and disabling functionality in GPU, which was done by @jane-intel 
https://github.com/openvinotoolkit/openvino/pull/12683#pullrequestreview-1088209308

So it's not possible to merge/fix it without turning on the functionality in GPU

**Comment by wilson-seok** [comment]
> > @OlehKravchyshyn Build error happened in build_rhel82_release. Please check it. image
> 
> @wilson-seok As I mentioned in mail this PR is connected with reverting and disabling functionality in GPU, which was done by @jane-intel [#12683 (review)](https://github.com/openvinotoolkit/openvino/pull/12683#pullrequestreview-1088209308)
> 
> So it has no sense to merge/fix it without turning on the functionality in GPU

Understood. Then can we convert this to draft until turning on EDDO in GPU plugin?

**Comment by tgubanova-lohika** [comment]
> > > @OlehKravchyshyn Build error happened in build_rhel82_release. Please check it. image
> > 
> > 
> > @wilson-seok As I mentioned in mail this PR is connected with reverting and disabling functionality in GPU, which was done by @jane-intel [#12683 (review)](https://github.com/openvinotoolkit/openvino/pull/12683#pullrequestreview-1088209308)
> > So it has no sense to merge/fix it without turning on the functionality in GPU
> 
> Understood. Then can we convert this to draft until turning on EDDO in GPU plugin?

I don't clearly understand how it should be done by process, as in fact this PR was not a draft and was completely ready for the review. From the development effort perspective the  task was fully implemented, but  it was broken externally. So I suppose some new task should be created to turn on the detectron itself and merge this PR together with it.

**Comment by wilson-seok** [comment]
> > > > @OlehKravchyshyn Build error happened in build_rhel82_release. Please check it. image
> > > 
> > > 
> > > @wilson-seok As I mentioned in mail this PR is connected with reverting and disabling functionality in GPU, which was done by @jane-intel [#12683 (review)](https://github.com/openvinotoolkit/openvino/pull/12683#pullrequestreview-1088209308)
> > > So it has no sense to merge/fix it without turning on the functionality in GPU
> > 
> > 
> > Understood. Then can we convert this to draft until turning on EDDO in GPU plugin?
> 
> I don't clearly understand how it should be done by process, as in fact this PR was not a draft and was completely ready for the review. From the development effort perspective the task was fully implemented, but it was broken externally. So I suppose some new task should be created to turn on the detectron itself and merge this PR together with it.

I just want to change this PR status as like pending. As you commented, let's pause now and will resume review and merge when EDDO in GPU is resolved.

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl [hunk 3]
```diff
@@ -133,21 +133,28 @@ KERNEL(eddo_ref_stage_0)
...
+        delta[1] = deltas[INPUT1_GET_INDEX(roi_idx, class_idx * 4 + 1, 0, 0)];
+        delta[2] = deltas[INPUT1_GET_INDEX(roi_idx, class_idx * 4 + 2, 0, 0)];
+        delta[3] = deltas[INPUT1_GET_INDEX(roi_idx, class_idx * 4 + 3, 0, 0)];
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl:156]: shouldn't here be like
INPUT1_GET_INDEX(roi_idx, class_idx , 1, 0)
INPUT1_GET_INDEX(roi_idx, class_idx , 2, 0)
INPUT1_GET_INDEX(roi_idx, class_idx , 3 ,0)

**Comment by wilson-seok** [comment]
@OlehKravchyshyn  Still we have build error in build_ubuntu18_release.
image

**Review by wilson-seok — APPROVED** [review]

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Please add this kernel in whitelist of program.cpp.

**Comment by tgubanova-lohika** [comment]
This Pr is dependent on 
[](https://github.com/openvinotoolkit/openvino/pull/13003 ) so PR 13003 should be merged first

File: src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl [hunk 4]
```diff
@@ -138,21 +138,28 @@ KERNEL(eddo_ref_stage_0)
...
-
         // width & height of box
         INPUT_TYPE2 box_size = (box.hi - box.lo + COORDINATE_OFFSET);
 
         // center location of box
         const INPUT_TYPE2 center = box.lo + HALF_ONE * box_size;
 
-        const INPUT_TYPE4 delta = vload4(offset, deltas) / DELTA_WEIGHTS;
+        INPUT_TYPE4 delta;
+        delta[0] = deltas[INPUT1_GET_INDEX(roi_idx, class_idx * 4, 0, 0)];
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl:158]: Why this is mdoified? vload4 will be better for performance?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: I think @yeonbok 's comment means separate the legacy code and the blocked format code using some jin constants as follows:
```c++
#ifdef USE_BLOCKED_FORMAT
        INPUT_TYPE4 res = vload4(ROI_COUNT * cls + idx, refined_boxes);
        output_boxes[idx0] = res[0];
        output_boxes[idx1] = res[1];
        output_boxes[idx2] = res[2];
        output_boxes[idx3] = res[3];
        output_scores[i] = score;
        output_classes[i] = cls;
#else
        vstore4(vload4(ROI_COUNT * cls + idx, refined_boxes), i, output_boxes);
        output_scores[idx_i4] = score;
        output_classes[idx_i5] = cls;
#endif
```

@yeonbok is it right?

Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: yes

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: @OlehKravchyshyn, please separate the legacy code and blocked format code

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: but despite these changes we haven't received performance regression, haven't we?

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: This code changes will affect blocked format as well as other formats.
Did you have performance regression test for other formats except blocked format too?

Comment (tgubanova-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: the code didn't support blocked format before, so as I know @wilson-seok  checked all the formats, here is comment from him
https://github.com/openvinotoolkit/openvino/pull/12561#issuecomment-1219479180

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: I just checked in8 model for blocked format. We need to check non blocked format impact on those 7 models which has EDDO op.
image

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: It is necessary to check regression, but I think it is also necessary to sperate block format code from legacy code for the ㅠbackward compatibility of non-blocked format.

Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: Even though the perf regression is not found from the test, the test is not exhaustive, so the original behavior is not expected to be changed if not necessary..

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: This model has performance regression. It's fp16 model. I'll share the model with @mhnap-lohika.
image

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: And this model has cldnn program build error. WIll share the model together.
image

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: @OlehKravchyshyn please separate code blocked format code from legacy codes using jit constant.

Comment (ahnyoung-paul) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: @wilson-seok, separation codes are updated. Could you please run regression test again?

Comment (wilson-seok) [reply src/plugins/intel_gpu/src/kernel_selector/core/cl_kernels/experimental_detectron_detection_output_ref.cl]: No regression in faster_rcnn_resnet50_fpn_coco.
cldnn program build error in instance-segmentation-security-0050 disappeared also.
http://ikswven004.ikor.intel.com:8080/job/i9-12900k-alderlake-05-fp16-32-dlbenchmark-compare-two-branches/54/artifact/regression/regression_public.txt/*view*/

**Review by ahnyoung-paul — APPROVED** [review]
LGTM

**Pull request merged**